drlhp/pref_interface.py

```diff
@@ -0,0 +1,213 @@
+import logging
+from functools import partial
+
+def handler(signum, frame):
+    print("Got no response, replaying segment video")
+    raise IOError("No response!")
+
+def fake_log(message, message_level, log_level, log_name):
+    if message_level >= log_level:
+        level_name = logging.getLevelName(log_level)
+        print(f"{level_name} - {log_name} - {message}")
+
+class PrefInterface:
+
+    def __init__(self, synthetic_prefs, max_segs, log_dir, zoom=4, channels=3,
+                 min_segments_to_test=2, max_idle_cycles=100, n_pause_frames=4,
+                 user_response_timeout=3):
+        if not synthetic_prefs:
+            self.vid_q = mp.get_context('spawn').Queue()
+            self.renderer = VideoRenderer(vid_queue=self.vid_q,
+                                          mode=VideoRenderer.play_through_mode,
+                                          zoom=zoom,
+                                          channels=channels)
+        else:
+            self.renderer = None
+        self.min_segments_to_test = min_segments_to_test
+        self.synthetic_prefs = synthetic_prefs
+        self.zoom = zoom
+        self.seg_idx = 0
+        self.segments = []
+        self.channels = channels
+        self.remaining_possible_pairs = 0
+        self.tested_pairs = set()  # For O(1) lookup
+        self.max_segs = max_segs
+        self.max_idle_cycles = max_idle_cycles
+        self.n_pause_frames = n_pause_frames
+        self.user_response_timeout = user_response_timeout
+        easy_tf_log.set_dir(log_dir)
+
+    def stop_renderer(self):
+        if self.renderer:
+            self.renderer.stop()
+
+    def run(self, seg_pipe, pref_pipe, remaining_pairs, kill_processes, log_level):
+        log_name = 'pref_interface.run'
+        pref_interface_fake_log = partial(fake_log, log_name=log_name, log_level=log_level)
+        self.recv_segments(seg_pipe)
+        idle_cycles = 0
+        while len(self.segments) < self.min_segments_to_test:
+            if kill_processes.value == 1:
+                pref_interface_fake_log(
+                    "Pref interface got kill signal, exiting (line 70)",
+                    logging.INFO)
+                return
+            #print(f"Pref interface only has {len(self.segments)} segments, waiting for {self.min_segments_to_test}, sleeping")
+            pref_interface_fake_log(f"Pref interface only has {len(self.segments)} segments, waiting for {self.min_segments_to_test}, sleeping", logging.DEBUG)
+            # This sleep time is load bearing, because if you sleep for too long you'll drop more segments on the ground due to
+            # not re-querying the segment pipe
+            time.sleep(0.05)
+            self.recv_segments(seg_pipe)
+
+        pref_interface_fake_log("Preference interface has more than two segments, starting to test", logging.INFO)
+        while True and kill_processes.value == 0:
+            seg_pair = None
+            while seg_pair is None:
```

Comment:
So, one reason why I could imagine it being good to keep things this way is that we actually have quite long environment reset times. We could be paused for something like 15-20 seconds with no new segments coming in. This is probably fine when we're doing human preferences, since humans aren't very fast and aren't going to get through very many preference comparisons that quickly, but I could imagine synthetic preferences doing a non-trivial amount of resampling of old pairs during those long periods of down time. Also, when segments are long, it can be pretty possible for the PrefInterface to sample a nontrivial number of duplicates during that time as well. It seems like a low number of duplicates is fine, but if we had a substantial it would make it harder to confidently test things like training size, if our effective training size is actually smaller. 

That said, this might only be a problem early in training, before the number of combinations gets so high that it wouldn't be possible to probabilistically "catch up" in the course of an env reset. I can try to empirically look at how frequently we hit "out of untested segments" as a measure of how much duplication we can ~roughly~ expect to do with waiting for 10 segments to start with

Reply:
Okay, I suspect this won't be a problem unless we start collecting very long segments relative to the length of episodes, especially if we change out the contents of self.segments relatively recently